File: dnaeuuon/TCT/음료수얼려먹기.cpp
#include <iostream>
#include <queue>
using namespace std;
int ice[1000][1000] = { 0, };
int visited[1000][1000] = { 0, };
int N, M;
int dx[4] = { 0,0,1,-1 };
int dy[4] = { 1,-1,0,0 };

void dfs(int x, int y) {
	visited[x][y] = 1;
	for (int i = 0; i < 4; i++) {
		int nx = x + dx[i];
		int ny = y + dy[i];
		if (nx < 0 || nx >= N || ny < 0 || ny >= M) {
			continue;
		}
		if (!visited[nx][ny] && (ice[nx][ny] == 0)) {
			dfs(nx, ny);
		}
	}
}
void bfs(int x, int y) {
	queue<int> q;
	q.push(x);
	q.push(y);
	ice[x][y]++;
	while (!q.empty()) {
		x = q.front();
		q.pop();
		y = q.front();
		q.pop();
		for (int i = 0; i < 4; i++) {
			int nx = x + dx[i];
			int ny = y + dy[i];
			if (nx >= 0 && nx < N && ny >= 0 && ny < M) {
				if (!ice[nx][ny]) {
					q.push(nx);
					q.push(ny);
					ice[nx][ny]++;
				}
			}
		}
	}
}
int main() {
	int input;
	cin >> N >> M; // За, ї­
	for (int i = 0; i < N; i++) {
		string str;
		cin >> str;
		for (int j = 0; j < M; j++) {
			ice[i][j] = str[j] - '0';
		}
	}
	int cnt = 0;
	for (int i = 0; i < N; i++) {
		for (int j = 0; j < M; j++) {
			if (!ice[i][j] && !visited[i][j]) {
				//dfs(i, j);
				bfs(i, j);
				cnt++;
			}
		}
	}
	cout << cnt << endl;
}

File: zoouniak/TCT/java/음료수_얼려먹기.java
package TCT.java;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;

public class 음료수_얼려먹기 {
    static boolean[][] ice;
    static boolean[][] visited;
    static int n, m;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] s = br.readLine().split(" ");
        n = Integer.parseInt(s[0]);
        m = Integer.parseInt(s[1]);
        ice = new boolean[n][m];
        visited = new boolean[n][m];
        for (int i = 0; i < n; i++) {
            String[] input = br.readLine().split("");
            for (int j = 0; j < m; j++) {
                if (Integer.parseInt(input[j]) == 1)
                    ice[i][j] = true;
            }

        }
        int result = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++)
                if (!ice[i][j] && !visited[i][j]) {
                    bfs(i, j);
                    result++;
                }
        }
        System.out.println(result);

    }

    private static void bfs(int i, int j) {
        int[] dx = {1, -1, 0, 0};
        int[] dy = {0, 0, -1, 1};
        Queue<int[]> q = new LinkedList<>();
        visited[i][j] = true;
        q.add(new int[]{i, j});
        while (!q.isEmpty()) {
            int x = q.peek()[0];
            int y = q.peek()[1];
            q.poll();
            for (int d = 0; d < 4; d++) {
                int nx = x + dx[d];
                int ny = y + dy[d];
                if (nx >= 0 && nx < n && ny >= 0 && ny < m) {
                    if (!ice[nx][ny] && !visited[nx][ny]) {
                        visited[nx][ny] = true;
                        q.add(new int[]{nx, ny});
                    }
                }
            }
        }
    }
}